src/transformers/models/llama/modeling_llama.py

```diff
@@ -590,15 +645,135 @@ def _upad_input(self, query_layer, key_layer, value_layer, padding_mask, query_l
+        else:
+            query_states = self.q_proj(hidden_states)
+            key_states = self.k_proj(hidden_states)
+            value_states = self.v_proj(hidden_states)
+
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        batch_size = query_states.shape[0]
+        query_length = query_states.shape[-2]
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value[0].shape[-2]
+        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+
+        if past_key_value is not None:
+            # reuse k, v, self_attention
+            key_states = torch.cat([past_key_value[0], key_states], dim=2)
+            value_states = torch.cat([past_key_value[1], value_states], dim=2)
+
+        past_key_value = (key_states, value_states) if use_cache else None
+
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+        # NOTE: Llama does not use dropout in the attention, hence the hard-coded
+        # dropout_p=0.0 independent of self.training.
+        # The batch_size = 1 case is handled separately to allow to dispatch on flash attention
+        # kernel.
+        # NOTE: As of PyTorch 2.1, SDPA can not dispatch to flash attention v2 in case an
+        # attention mask passed. Possible solution: use nested tensors.
+
+        # Below, we ignore the attention mask in some cases to allow the dispatch to flash attention.
+        if batch_size == 1 and padding_mask is None:
+            if query_length == 1:
+                is_causal = False
+                attn_mask = None
+            elif kv_seq_len == query_length:
+                is_causal = True
+                attn_mask = None
+            else:
+                # Unfortunately, for query_length > 1, we can not generally ignore the attention mask, as SDPA causal mask generation
+                # may be wrong. Reference: https://github.com/pytorch/pytorch/issues/108108
+                attn_mask = torch.max(attention_mask, torch.tensor(torch.finfo(key_states.dtype).min))
+                is_causal = False
+        elif attention_mask is not None:
+            attn_mask = torch.max(attention_mask, torch.tensor(torch.finfo(key_states.dtype).min))
+            is_causal = False
+        else:
+            attn_mask = None
+            is_causal = True
+
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
+            query_states, key_states, value_states, attn_mask=attn_mask, dropout_p=0.0, is_causal=is_causal
+        )
```

Comment:
Unless I am mistaken I think that this could be wrapped in a method of `LLamaAttention` and have a simple method `sdpa_forward()` and have a simple check `if_sdpa_available()` that calls that method and call the vanilla attention instead of having duplicate classes, if and only if it is the only required change to enable SDPA. 

1- Classes that use `LlamaAttention` will directly benefit from SDPA without introducing an extra logic in the init with just `make fix-copies`
2- If tomorrow we need to introduce a fix for `pretraining_tp` or anything related with attention we need to check that the fix has been properly propagated into all `xxxAttention` and `xxxSDPAAttention` which IMO makes it hard to maintain and cross check. Having everything in `xxxAttention` will automatically propagate the fix for SDPA and vanilla attention.
3- The causal mask expected by SDPA looks similar than the causal mask expected by the vanilla attention
4- Again, as a user I would be very surprised to see that my `xxxAttention` class has been suddenly changed to `xxxSDPAAttention` class with no explicit intervention.

For reference here is the SDPA implementation of Falcon: https://github.com/huggingface/transformers/blob/main/src/transformers/models/falcon/modeling_falcon.py#L476-L495 - 5 LoC for exactly the same result. 

For the reasons stated above I would advocate to modify the `LlamaAttention` rather than introducing a new class `LlamaSDPAAttention`

Reply:
1. Yes. But we have a `#Copied from` logic designed for this.
2. I don't follow, if we split flash and not flash, long term maintenance is split and  the code is "simpler". I don't see why `pretraining_tp` would change; but I would certainly see why our cache logic could change or torch's attention mask requirements (like they did with torch.where on bool) could evolve. 
3. Cool if indeed true and we only have 1 line like Falcon.
4. I think that at least is a **visible** change, vs a silent change in the forward 😅 I'm not sure how well support this operation is by torch compile, ONNX etc
5. Computing the `torch.max` in the forward is not optimal, if we need to cache this after computing it once, we are polluting the class)

IMO adding if else in the forward that changes the behaviour of the model is what we have been trying to avoid (but also prevent) in the library. 

Now let's stay open! Leaving this if this does not add unnecessary changes for other users and stays readable cool (they are not all using pytorch 2.0). 

My effort in adding NLP model was to standardize  and simplify our attention layer with just what's needed to keep the code clean, hoping to keep doing this!

Reply:
About the `torch.max`: it is a legacy from https://github.com/huggingface/transformers/blob/v4.30.2/src/transformers/models/llama/modeling_llama.py#L226, apparently it has been removed in https://github.com/huggingface/transformers/pull/24891, I indeed noticed that we have some `- inf` in the mask now due to both `_make_causal_mask` and `_expand_mask` masking the same token. I will remove the `torch.max` given that this behavior of avoiding `-inf` in the mask has changed.

About the change of default class from `LlamaAttention` to `LlamaSDPAAttention`, it sounds fine to me given that `LlamaSDPAAttention` inherits from `LlamaAttention`.

Reply:
The simplest and more explicit change here would be to have the `LlamaSDPAAttention` defined as it is currently done. 

This is more future-proof for future possible attention layers that we may have to add (like the `LlamaFlashAttention` layer), and the attention layer used under the hood is very clear to the end user.

IMO it also makes the code less error-prone as we propagate this change to several other architectures.